### DISCUSSION TITLE
PICO W + GRBLHAL + CNCSHIELDV3 = 4 Axis hotwire :-0

### DISCUSSION THREAD

**Discussion by flyinggazzad4** [terjeio/ioSender#380] [General]
Hi Can I start by saying what a fantastic repository you have set up here! I'm really very new to GitHub in fact this is my first attempt at a discussion. As the title suggests I' looking at a very low tech/low cost solution to drive a 4 independent  axis wire cutter fabricated out of old ender 3 printer using NEMA 17 stepper motors and a DC PWM motor driver to regulate the temperature output of the ni-chrome wire. the DC motor driver has a 0-10K pot to set 0-100% output so in time may be it is possible to incorporate into one software solution but for now I'm looking at harnessing the power of a Raspberry Pi PICO-W and a 4 way CNC shield originally used on the Arduino UNO platform. From the limited research I've done it is possible to operate all 4 stepper motors independently using the Arduino UNO just not in conjunction with GRBL. In web searching for a solution I was fortunate enough to find GRBLHAL, which I think is the the answer I'm looking for! - From reading the threads I have a level of confidence that the PICO is capable of driving this setup? - I looked at the web builder and can select PICO from the menu to create UF2 file to download onto the PICO, and potentially WIFI in the future! however not to get distracted from the core activity -  I'm looking to see how I can determine the pinout details for the PICO - I have a very basic understanding of coding (learnt from modifying 3D marlin config.h files, so armed with this information and what i think would be my best shot at achieving his goal would be opting for the Generic_map_4axis.h  file in VS code i can see reference to GPIO pins for some of the outputs? they look differently arranged to the only pinout diagram I've identified as copied below.

![thumbnail_105925371-dda26b00-5ff4-11eb-8465-9f80e4bcfc91](https://github.com/terjeio/ioSender/assets/167357866/c679e3d2-8b52-48c6-98f7-97a8370d7ddd)

Vs what I think I can deduce from the generic_map_4axis

Function		                     GPIO PIN assignment

AUXOUTPUT0_PIN               5

X_DIRECTION_PIN                6	
Y_DIRECTION_PIN                7	
Z_DIRECTION_PIN                8	
AUXOUTPUT1_PIN               9           Assume this would be axis 4

STEPPERS_ENABLE_PIN     10	

X_LIMIT_PIN                       11	
Y_LIMIT_PIN                       12	
Z_LIMIT_PIN                       13	
AUXINPUT0_PIN                14             Assume this would be axis 4 limit switch
 
SPINDLE_ENABLE_PIN       15	

COOLANT_FLOOD_PIN     16	
COOLANT_MIST_PIN        17

RESET_PIN                        18	
FEED_HOLD_PIN               19	
CYCLE_START_PIN             20	

AUXINPUT1_PIN               21
AUXINPUT3_PIN               22

SPINDLE_ENABLE_PIN      26
SPINDLE_DIR_PIN             27
AUXINPUT2_PIN              28

I don't see the Step increment Pin assignments? - It  is likely that I just do not understand how to interpret the code structure sufficiently well.

Is is simply 1,2,3,4? - I read in the threads that PIO needed consecutive assignment of GPIO pins.

![gpiopins](https://github.com/terjeio/ioSender/assets/167357866/2e2ebf44-a76b-4701-a15e-f6d88e21d83d)

It feels like I'm so tangibly close to making this a reality but I'm stumped at is seemingly last hurdle - I appreciate you will likely have lots of people reaching out to you for advice and any commentary you would be happy to share or input you might have for me would be very gratefully appreciated

with best regards

Gareth

**Comment by terjeio** [comment]
> I don't see the Step increment Pin assignments? 

Not sure what you mean by this, I interpret it as the increment is 1:

`STEP_PINS_BASE` is the X step output (gpio 2), Y step will be `STEP_PINS_BASE + 1` (gpio 3) and so on.

If you flash the firmware you can use the `$pins` command to get a list of how the pins are assigned in the selected configuration.

FYI for grblHAL gpio 0 and 1 is typically reserved for a UART channel (serial port).

**Reply by flyinggazzad4** [reply]
Hi Terjeio, Thank you for confirming  I'm very finding my feet and being self taught I had not before seen this notation and wanted confirmation that I was understanding correctly. I do appreciate you responding to me its my first time using the GitHub platform I come from a mechanical background rather then software. with kind regards

Gareth

**Reply by phil-barrett** [reply]
Note that pinout only supports 3 Axes/motors.  Step pins are 2,3,4.  AuxOutput0 is 5.  You will need to move it elsewhere to get 5 as the 4th axis/motor pin.

**Reply by terjeio** [reply]
_generic_map_4axis.h_ do support 4 axes - when 3 are configured the 4th axis pins are [assigned as aux i/o](https://github.com/grblHAL/RP2040/blob/f34e564f4eb170b9f1c1ca58263e269f989d3735/generic_map_4axis.h#L53-L64). However there are duplicate definitions of aux pins if configured for 3 and the PWM spindle is not enabled - I'll have to fix that.

**Reply by flyinggazzad4** [reply]
Good evening  Terjeio hope you are well I have a quick question,  I see in the web builder there is a CORE X-Y kinematic option - I see many advantages to CORE XY in the design im using and would like to investigate further this type of setup, I would also require the Z-A axis set to behave in effect as a second X1-Y1 setup independent of the first X-Y axis set the XY will map the root profile whilst Z-A follows the the top profile. is this at all possible within a GRBLHAL PICO setup? - Having good results thus far with trials thank you! -

With best regards

Gareth

**Reply by terjeio** [reply]
>  is this at all possible within a GRBLHAL PICO setup?

If I understand you correctly then I think so, you will have to add a new kinematics implementation for it.

**Comment by flyinggazzad4** [comment]
hi Terjeio

thank you for responding, im not sure i would have the skill set required to apply a new kinematics implementation- i would love to but also respectful of my limitations, is it a relatively simple task?

with best regards


gareth

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
